src/res/mod.rs

```diff
@@ -244,124 +257,295 @@ impl Resources {
+    /// in case the resource does not exist.
+    pub fn try_fetch<T, P>(&self) -> Option<Fetch<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            Fetch {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| Fetch {
                 inner: r.borrow(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
     /// Fetches the resource with the specified type `T` mutably.
     ///
     /// Please see `fetch` for details.
-    pub fn fetch_mut<T>(&self) -> FetchMut<T>
+    pub fn fetch_mut<T, F, P>(&self, with: F) -> FetchMut<T, P>
     where
         T: Resource,
+        F: FnOnce() -> T,
     {
-        self.try_fetch_mut().expect(RESOURCE_NOT_FOUND)
+        let res_id = ResourceId::new::<T>();
+        let r = self.resources
+            .get(&res_id)
+            .unwrap_or_else(|| self.def_res::<T>(res_id, with()));
+
+        FetchMut {
+            inner: r.borrow_mut(),
+            phantom: PhantomData,
+        }
     }
 
-    /// Like `fetch_mut`, but returns an `Option` instead of panicking in the case of the resource
-    /// being accessed mutably.
-    pub fn try_fetch_mut<T>(&self) -> Option<FetchMut<T>>
+    /// Like `fetch_mut`, but returns an `Option` instead of inserting a default value
+    /// in case the resource does not exist.
+    pub fn try_fetch_mut<T, P>(&self) -> Option<FetchMut<T, P>>
     where
         T: Resource,
     {
-        self.try_fetch_internal(TypeId::of::<T>()).map(|r| {
-            FetchMut {
+        let res_id = ResourceId::new::<T>();
+
+        self.resources
+            .get(&res_id)
+            .or_else(|| self.get_res::<T>(&res_id))
+            .map(|r| FetchMut {
                 inner: r.borrow_mut(),
                 phantom: PhantomData,
-            }
-        })
+            })
     }
 
-    fn try_fetch_internal(&self, id: TypeId) -> Option<&TrustCell<Box<Resource>>> {
-        self.resources.get(&ResourceId(id))
+    fn def_res<T>(&self, res_id: ResourceId, value: T) -> &TrustCell<Box<Resource>>
+    where
+        T: Resource,
+    {
+        let mut new = self.new.lock();
+        let b: &mut Box<_> = new.entry(res_id)
+            .or_insert_with(|| Box::new(TrustCell::new(Box::new(value) as Box<Resource>)));
+
+        Resources::box_to_cell_ref(b)
+    }
+
+    fn get_res<T>(&self, res_id: &ResourceId) -> Option<&TrustCell<Box<Resource>>>
+    where
+        T: Resource,
+    {
+        let new = self.new.lock();
+        let b: &Box<_> = match new.get(&res_id) {
+            Some(x) => x,
+            None => return None,
+        };
+
+        Some(Resources::box_to_cell_ref(b))
+    }
+
+    /// May only be called by `def_res` or `get_res`.
+    fn box_to_cell_ref<'a, 'b>(
```

Comment:
I am assuming you missed it because you didn't comment on it in the last chain.
Couldn't you make this safer by taking `&self` as parameter and connecting the returned references life time to it?